### PR TITLE
Refactorized session management and removed recursion

### PR DESCRIPTION
**Pull request by charan7386**
**Changes Made:**
Used [Transitions Library](https://pypi.org/project/transitions/), in future the menu may become lengthy, so this library will be helpful in viewing states and transition visually.
Removed recursion

**Fixes/Issues:**
fixed issue #149

### CODE REVIEW

File: wplay/utils/session_manager.py [hunk 1]
```diff
@@ -10,137 +12,171 @@
...
+                self.username = self.answers_menu['restore']
+                self.save_session = True
+                return True
+
+        # Handle when person choose 'Create a new session'
+        elif self.answers_menu['user_options'] == self.user_options['save']:
+            self.username = self.answers_menu['save']
+            self.save_session = True
+            return self.__verify_if_session_file_exists()
+
```
Comment (rpotter12) [inline wplay/utils/session_manager.py]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5097332)

**Comment by rpotter12** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 6
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5097332&bid=16767862)

File: wplay/utils/session_manager.py [hunk 2]
```diff
@@ -10,137 +12,171 @@
...
+        # Handle when person choose 'Continue without saving'
+        elif self.answers_menu['user_options'] == self.user_options['continue']:
+            self.username, self.save_session = None, False
+            return True
+
+        # Handle when person choose 'Delete a session'
+        elif self.answers_menu['user_options'] == self.user_options['delete']:
+            if len(self.answers_menu['delete']) > 0:
+                [self.__delete_session_data(user_data_folder_path / username) for username in self.answers_menu['delete']]
```
Comment (rpotter12) [inline wplay/utils/session_manager.py]: Codacy Issue found: [Expression "[self.__delete_session_data((user_data_folder_path) / (username)) for username in self.answers_menu['delete']]" is assigned to nothing](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5097332)

File: wplay/utils/session_manager.py [hunk 3]
```diff
@@ -10,137 +12,171 @@
...
+            {
+                'type': 'confirm',
+                'name': 'overwrite_data',
+                'message': 'There is already a session with that name, overwrite it?',
+                'default': True
+            }
+        ]
+    
+    def get_answer_menu(self):
+        self.answers_menu = prompt(self.questions_menu, style=menu_style)
```
Comment (rpotter12) [inline wplay/utils/session_manager.py]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5097332)

File: wplay/utils/session_manager.py [hunk 4]
```diff
@@ -10,137 +12,171 @@
...
-
-    shutil.rmtree(path, onerror=handleError)
-
-# endregion
+def session_manager():
+    done = False
+    obj = CliWhatsappPlay()
+    while(not done):
+        if(obj.questions_menu != None): obj.reset_fields()
+        machine = Machine(obj, states, transitions=transitions, initial='start')
```
Comment (rpotter12) [inline wplay/utils/session_manager.py]: Codacy Issue found: [Unused variable 'machine'](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5097332)

File: wplay/utils/session_manager.py [hunk 5]
```diff
@@ -10,137 +12,171 @@
...
+    def reset_fields(self):
+        self.data_filenames = None
+        self.questions_menu = None
+        self.question_overwrite = None
+        self.answers_menu = None
+        self.username = None
+        self.save_session = False
+
+
+    def create_user_data_folder(self):
```
Comment (rpotter12) [inline wplay/utils/session_manager.py]: Codacy Issue found: [Method could be a function](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5097332)

File: wplay/utils/session_manager.py [hunk 6]
```diff
@@ -10,137 +12,171 @@
...
+    def __verify_if_session_file_exists(self):
+        if self.username in self.data_filenames:
+            answer_overwrite = prompt(self.question_overwrite, style=menu_style)
+            if answer_overwrite['overwrite_data']:
+                self.__delete_session_data(user_data_folder_path / self.username)
+            return answer_overwrite['overwrite_data']
+        return True
+
+
+    def __delete_session_data(self, path):
```
Comment (rpotter12) [inline wplay/utils/session_manager.py]: Codacy Issue found: [Method could be a function](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5097332)

**Comment by priyansh19** [comment]
@charan7386 please update your repo from upstream repo :+1:

**Comment by xandao6** [comment]
It's okay with his repository.

**Comment by rpotter12** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 6
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5097332&bid=16767862)

File: wplay/utils/session_manager.py [hunk 7]
```diff
@@ -1,146 +1,182 @@
...
```
Comment (rpotter12) [inline wplay/utils/session_manager.py]: Codacy Issue found: [Method could be a function](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5097332)

File: wplay/utils/session_manager.py [hunk 8]
```diff
@@ -1,146 +1,182 @@
...
+
+        # Handle when person choose 'Continue without saving'
+        elif self.answers_menu['user_options'] == self.user_options['continue']:
+            self.username, self.save_session = None, False
+            return True
+
+        # Handle when person choose 'Delete a session'
+        elif self.answers_menu['user_options'] == self.user_options['delete']:
+            if len(self.answers_menu['delete']) > 0:
+                [self.__delete_session_data(user_data_folder_path / username) for username in self.answers_menu['delete']]
```
Comment (rpotter12) [inline wplay/utils/session_manager.py]: Codacy Issue found: [Expression "[self.__delete_session_data((user_data_folder_path) / (username)) for username in self.answers_menu['delete']]" is assigned to nothing](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5097332)

File: wplay/utils/session_manager.py [hunk 9]
```diff
@@ -1,146 +1,182 @@
...
-
-    shutil.rmtree(path, onerror=handleError)
-
-# endregion
+def session_manager():
+    done = False
+    obj = CliWhatsappPlay()
+    while(not done):
+        if(obj.questions_menu != None): obj.reset_fields()
+        machine = Machine(obj, states, transitions=transitions, initial='start')
```
Comment (rpotter12) [inline wplay/utils/session_manager.py]: Codacy Issue found: [Unused variable 'machine'](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5097332)

File: wplay/utils/session_manager.py [hunk 10]
```diff
@@ -1,146 +1,182 @@
...
+    def reset_fields(self):
+        self.data_filenames = None
+        self.questions_menu = None
+        self.question_overwrite = None
+        self.answers_menu = None
+        self.username = None
+        self.save_session = False
+
+
+    def create_user_data_folder(self):
```
Comment (rpotter12) [inline wplay/utils/session_manager.py]: Codacy Issue found: [Method could be a function](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5097332)

File: wplay/utils/session_manager.py [hunk 11]
```diff
@@ -1,146 +1,182 @@
...
+                self.username = self.answers_menu['restore']
+                self.save_session = True
+                return True
+
+        # Handle when person choose 'Create a new session'
+        elif self.answers_menu['user_options'] == self.user_options['save']:
+            self.username = self.answers_menu['save']
+            self.save_session = True
+            return self.__verify_if_session_file_exists()
+
```
Comment (rpotter12) [inline wplay/utils/session_manager.py]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5097332)

File: wplay/utils/session_manager.py [hunk 12]
```diff
@@ -1,146 +1,182 @@
...
+            {
+                'type': 'confirm',
+                'name': 'overwrite_data',
+                'message': 'There is already a session with that name, overwrite it?',
+                'default': True
+            }
+        ]
+    
+    def get_answer_menu(self):
+        self.answers_menu = prompt(self.questions_menu, style=menu_style)
```
Comment (rpotter12) [inline wplay/utils/session_manager.py]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5097332)

**Comment by vaishnavi-janardhan** [comment]
@xandao6 Can you please modify the label `GSSoC 2020` to `gssoc20`

**Comment by xandao6** [comment]
@vaishnavi-janardhan Ok

**Pull request merged**